Add VictorOps Notification Support

ichekrygin opened the pull request:
### Overview

VictorOps is a hub for centralizing the flow of information throughout the incident lifecycle. Driven by IT and DevOps system data, VictorOps provides a unified platform for real-time alerting, collaboration, and documentation (https://victorops.com/)

This PR adds functionality to `POST` alert notifications to VictorOps via [REST API](http://victorops.force.com/knowledgebase/articles/Integration/Alert-Ingestion-API-Documentation/)

Feature Request [#6411](https://github.com/grafana/grafana/issues/6411)

![image](https://cloud.githubusercontent.com/assets/324803/19783227/ee7a192c-9c45-11e6-9320-752f5f3c62ed.png)

IllyaChekrygin committed:
Add VictorOps alert notification capability

IllyaChekrygin committed:
Fix user-facing text

utkarshcmu commented on public/app/features/alerting/alert_tab_ctrl.ts:
```diff
@@ -87,6 +87,7 @@ export class AlertTabCtrl {
     switch (type) {
       case "email": return "fa fa-envelope";
       case "slack": return "fa fa-slack";
+      case "victorops": return "fa fa-exclamation-triangle";
```
I don't think using `fa-exclamation-triangle` is a good idea as usually `fa-exclamation-triangle` signifies error. May be you can use something else. Nearest to what I found to VictorOps logo was `fa-pagelines`.

IllyaChekrygin committed:
Change VictorOps fa from `fa-exclamation-triangle` -> `fa-pagelines`

utkarshcmu commented on pkg/services/alerting/notifiers/victorops.go:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	messageType := evalContext.Rule.State
+	if evalContext.Rule.State == models.AlertStateAlerting { // translate 'Alerting' to 'CRITICAL' (Victorops analog)
```
Some concerns:
1. What if someone wants a notification to be sent to VictorOps on "No Data" or "Exec Error"? How should we handle those conditions?
2. What if someone wants to auto-resolve on "OK" state?

I am writing a notifier for PagerDuty so thinking these questions are coming into my mind. :)

utkarshcmu replied:
I am just thinking how can we make victorOps notifier more complete handling most of the use cases. :)

utkarshcmu replied:
Fyi, Check the screenshot in https://github.com/grafana/grafana/issues/6291

ichekrygin replied:
@utkarshcmu: I like the idea of providing options in handling `ExecutionError`, I guess mainly to suppress the noise, although as a temporary measure. I need to test this w/ VictorOps since I am not exactly sure how would it handle POST on `ExecutionError` (i.e. will it trigger an incident or will it just post to the timeline. I tend to think former, since this seems to be a default behavior)

I think the handling of `NoData` is slightly different since it is a result of conditional check `If no data points or all values are null`, hence, the user can either set it to: `OK`, `Alerting`, `KeepLast` or `NoData`. One potential use case I see if the user wants to trigger `Warning` type notification or possibly just posting to `VictorOps` timeline on this event.

As for auto-resolving on `OK` state, this is already handled by `VictorOps` by default. I ran a couple of tests w/ current code, and `VictorOps` recognized the `Recovery` state and clear the incidents.

utkarshcmu replied:
I just wanted to bring up these questions as I am facing the same questions while writing the code for pagerduty integration. Its up to you and the maintainers of Grafana, how you guys want to implement it for Grafana v4.0. 😊

I have got a second opinion now and I will remove the "auto-resolve" for pagerduty for at least Grafana v4.0. Will wait for the feedback from the community.

utkarshcmu replied:
Thanks @ichekrygin for sharing the feedback.

IllyaChekrygin committed:
Add option to alert on ExecutionError

IllyaChekrygin committed:
Fix return on `nil` and cleanup debug messages

IllyaChekrygin committed:
Resolve rebase w/ upstream conflicts.
Remove: `Alert on Exec Error` (no longer needed)
Add: `ruleUrl` to VictorOps message body.

IllyaChekrygin committed:
Add VictorOps Test.

IllyaChekrygin committed:
Fix formatting

ichekrygin commented:
@bergquist @utkarshcmu @torkelo Hi guys, could you please git this PR another look?

utkarshcmu commented:
I can test this tonight! :)

utkarshcmu commented:
I tested the PR and it works. It creates an incident in VictorOps on state = Alerting and then resolves it when state = OK. If anyone from @bergquist or @torkelo wants to test the PR, they can use my trial VictorOps account, which will expire in next 12 days. 👍

wagner-victorops commented:
You can find your REST endpoint URL in VictorOps at _Settings >> Alert Behavior >> Integrations >> Legacy REST._

nategrieb commented on pkg/services/alerting/notifiers/victorops.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	messageType := evalContext.Rule.State
+	if evalContext.Rule.State == models.AlertStateAlerting { // translate 'Alerting' to 'CRITICAL' (Victorops analog)
```
the "/n" in the state_message field (line 92) causes invalid JSON to be sent through. Can this be removed? and replaced with a " " or a " - " ?
